File: src/phiid/phiid_pid_comparison.m
clear all;
close all;
clc;

cd '/media/nadinespy/NewVolume/my_stuff/work/PhD/my_projects/PhiIDComparison'
addpath '/media/nadinespy/NewVolume/my_stuff/work/PhD/my_projects/PhiIDComparison/scripts'
addpath '/media/nadinespy/NewVolume/my_stuff/work/toolboxes_matlab'
% addpath '/media/nadinespy/NewVolume/my_stuff/work/PhD/my_projects/PhiIDComparison/scripts/heatmaps'

PATHOUT = '/media/nadinespy/NewVolume/my_stuff/work/PhD/my_projects/PhiIDComparison/results/';
javaaddpath('infodynamics.jar');

%% calculating information atoms

% System size, coupling matrix, and vector of noise values
nvar = 2;
npoints = 2000;
tau = 1;
error_vec   = linspace(0.01, 0.99, 100);
coupling_vec = linspace(0.01,0.49, 100);

phiid_all_err_coup_mmi = zeros(16, size(coupling_vec,2), size(error_vec, 2));
phiid_all_err_coup_ccs = zeros(16, size(coupling_vec,2), size(error_vec, 2));


for i = 1:length(coupling_vec)
	A = coupling_vec(i)*ones(nvar);
	
	for j = 1:length(error_vec)
		err = error_vec(j);
		spectral_radius = max(abs(eig(A)));
		%X = statdata_corr_errors(A, npoints, err);
		X = statdata_corr_errors2(A, npoints, err);
		% X = statdata_corr_errors3(A, npoints, err);
		phiid_all_err_coup_mmi(:,i,j) = struct2array(PhiIDFull(X, tau, 'MMI'))';
		phiid_all_err_coup_ccs(:,i,j) = struct2array(PhiIDFull(X, tau, 'ccs'))';
		
	end 
	i
	%spectral_radius
end 

% save([PATHOUT 'phiid_all_err_coup_ccs_rtr.mat'],'phiid_all_err_coup_ccs_rtr');
% save([PATHOUT 'phiid_all_err_coup_ccs_sts.mat'],'phiid_all_err_coup_ccs_sts');
% save([PATHOUT 'phiid_all_err_coup_mmi_rtr.mat'],'phiid_all_err_coup_ccs_mmi');
% save([PATHOUT 'phiid_all_err_coup_mmi_sts.mat'],'phiid_all_err_coup_ccs_mmi');

%% double-redundancy & double-synergy

% extract double-redundancy & synergy term
% rows: couplings; errors: columns
phiid_all_err_coup_ccs_rtr = squeeze(phiid_all_err_coup_ccs(1,:,:));			% coupling vector will now be in the rows, and error vector in the columns
phiid_all_err_coup_ccs_sts = squeeze(phiid_all_err_coup_ccs(16,:,:));

phiid_all_err_coup_mmi_rtr = squeeze(phiid_all_err_coup_mmi(1,:,:));
phiid_all_err_coup_mmi_sts = squeeze(phiid_all_err_coup_mmi(16,:,:));

% heatmaps
x_axis = {'0.01', '', '', '', '', '', '', '', '', '0.1', '', '', '', '', '', '', '', '', '', '0.2', '', '', '', '', '', '', '', '', '', '0.3', '', '', '', '', '', '', '', '', '', ... 
	'0.4', '', '', '', '', '', '', '', '', '', '0.5', '', '', '', '', '', '', '', '', '', '0.6', '', '', '', '', '', '', '', '', '', '0.7', '', '', '', '', '', '', '', '', '', ... 
	'0.8', '', '', '', '', '', '', '', '', '', '0.9', '', '', '', '', '', '', '', '', '0.99', ''}; 
y_axis = {'0.01', '', '', '', '', '', '', '', '', '',  '', '', '', '', '', '', '', '',  '', '0.1',  '', '', '', '', '', '', '', '', '', '',  '', '', '', '', '', '', '', '', '', ... 
	'0.2', '', '', '', '', '', '', '', '', '',  '', '', '', '', '', '', '', '', '',  '', '0.3', '', '', '', '', '', '', '', '', '',  '', '', '', '', '', '', '', '', '', '', ... 
	'0.4', '', '', '', '', '', '', '', '', '',  '', '', '', '', '', '', '', '', '', '',  '0.5'};

% using matlab built-in function for heatmaps:
figure;
clf
h = heatmap(phiid_all_err_coup_mmi_rtr, 'Colormap', parula, 'ColorbarVisible', 'on') ;
h.YDisplayLabels = y_axis;
h.XDisplayLabels = x_axis;
h.XLabel = 'noise correlation';
h.YLabel = 'coupling strength';
title('double-redundancy mmi');
exportgraphics(gcf, [PATHOUT '2node_phiid_all_err_coup_mmi_rtr2.png']);

figure;
clf
h = heatmap(phiid_all_err_coup_mmi_sts, 'Colormap', parula, 'ColorbarVisible', 'on') ;
h.YDisplayLabels = y_axis;
h.XDisplayLabels = x_axis;
h.XLabel = 'noise correlation';
h.YLabel = 'coupling strength';
title('double-synergy mmi');
exportgraphics(gcf, [PATHOUT '2node_phiid_all_err_coup_mmi_sts2.png']);

figure;
clf
h = heatmap(phiid_all_err_coup_ccs_rtr, 'Colormap', parula, 'ColorbarVisible', 'on') ;
h.YDisplayLabels = y_axis;
h.XDisplayLabels = x_axis;
h.XLabel = 'noise correlation';
h.YLabel = 'coupling strength';
title('double-redundancy ccs');
exportgraphics(gcf, [PATHOUT '2node_phiid_all_err_coup_ccs_rtr2.png']);

figure;
clf
h = heatmap(phiid_all_err_coup_ccs_sts, 'Colormap', parula, 'ColorbarVisible', 'on') ;
h.YDisplayLabels = y_axis;
h.XDisplayLabels = x_axis;
h.XLabel = 'noise correlation';
h.YLabel = 'coupling strength';
title('double-synergy ccs');
exportgraphics(gcf, [PATHOUT '2node_phiid_all_err_coup_ccs_sts2.png']);

close all;

%load('A2b.mat')
%npoints = 2000;
%X = statdata(A2b,npoints);

% heatmaps (using "customizable" heatmaps: https://uk.mathworks.com/matlabcentral/fileexchange/24253-customizable-heat-maps,
% need to extract respective files from zip)
% figure;
% clf
% h = heatmap(phiid_all_err_coup_mmi_rtr, x_axis, y_axis, [], 'TickAngle', 0);
% colormap(parula);

% phiid_atom_names = {'rtr',  'rtx', 'rty', 'rts', 'xtr', 'xtx', 'xty', 'xts', 'ytr', 'ytx', 'yty', 'yts', 'str', 'stx', 'sty', 'sts'};

File: src/phiid/phiid_pid_comparison.py
# -*- coding: utf-8 -*-
"""
Spyder Editor

This is a temporary script file.
"""

%reset

from oct2py import octave as oc
import seaborn as sns
import numpy as np
import scipy.io
import os

PATHIN = '/media/nadinespy/NewVolume/my_stuff/work/PhD/my_projects/PhiIDComparison/scripts/'
PATHOUT = '/media/nadinespy/NewVolume/my_stuff/work/PhD/my_projects/PhiIDComparison/results/'


os.chdir('/media/nadinespy/NewVolume/my_stuff/work/PhD/my_projects/PhiIDComparison/scripts/')
oc.javaaddpath('/media/nadinespy/NewVolume/my_stuff/work/PhD/my_projects/PhiIDComparison/scripts/infodynamics.jar');
oc.addpath(oc.genpath('/media/nadinespy/NewVolume/my_stuff/work/PhD/my_projects/PhiIDComparison/'))
oc.addpath('/media/nadinespy/NewVolume/my_stuff/work/toolboxes_matlab/')
oc.eval('pkg load statistics')                                                  # needs to be loaded, as otherwise normrnd() in statdata will not be recognized



# HIER ABSOLUTEN PFAD MACHEN <3
oc.load(['/media/nadinespy/NewVolume/my_stuff/work/PhD/my_projects/PhiIDComparison/scripts/A2b.mat'])

nvar = 2
npoints = 10000
tau = 1
error_vec   = np.linspace(0.01, 0.99, 100)
coupling_vec = np.linspace(0.01,0.49, 100)

phiid_all_err_coup_mmi = np.zeros((16, len(coupling_vec), len(error_vec)))
phiid_all_err_coup_ccs = np.zeros((16, len(coupling_vec), len(error_vec)))


for i in range (1, len(coupling_vec)):
	A = coupling_vec[i]*np.ones(shape=(nvar,nvar))
	
	for j in range(1, len(error_vec)):
		err = error_vec[j]
		X = oc.statdata_corr_errors(A, npoints, err)                                  # for whatever reason, the rng() in statdata_corr_error.m doesn't work when called via oct2py (for now, it's disabled)
		phiid_all_err_coup_mmi[:,i,j] = struct2array(oc.PhiIDFull(X, tau, 'MMI')).T   # PhiIDFull.m doesn't work either, as mvnpdf() is not recognized
		phiid_all_err_coup_ccs[:,i,j] = struct2array(oc.PhiIDFull(X, tau, 'ccs')).T   # struct2array is from matlab - how to deal with struct outputs in python?
	i	

phiid_all_err_coup_mmi_rtr = np.squeeze(phiid_all_err_coup_mmi[1,:,:])
phiid_all_err_coup_mmi_sts = np.squeeze(phiid_all_err_coup_mmi[16,:,:])

os.chdir('/media/nadinespy/NewVolume/my_stuff/work/PhD/my_projects/PhiIDComparison/results/')

np.save(phiid_all_err_coup_ccs, phiid_all_err_coup_ccs)
np.save(phiid_all_err_coup_mmi, phiid_all_err_coup_mmi)


phiid_all_err_coup_ccs = scipy.io.loadmat('phiid_all_err_coup_ccs.mat',squeeze_me=True)['phiid_all_err_coup_ccs']              # loads file as a dictionary where 4th row is the actual data
phiid_all_err_coup_mmi = scipy.io.loadmat('phiid_all_err_coup_mmi.mat',squeeze_me=True)['phiid_all_err_coup_mmi']    


#phiid_all_err_coup_ccs = list(phiid_all_err_coup_ccs.values())[3]                   # put all values from dict into a list, and take 3rd entry (which has the data)
#phiid_all_err_coup_mmi = list(phiid_all_err_coup_mmi.values())[3]

phiid_all_err_coup_mmi_rtr = np.squeeze(phiid_all_err_coup_mmi[0,:,:])
phiid_all_err_coup_mmi_sts = np.squeeze(phiid_all_err_coup_mmi[15,:,:])

#heatmap
sns.set_theme()

ax = sns.heatmap(phiid_all_err_coup_mmi_rtr)
ax1 = sns.heatmap(phiid_all_err_coup_mmi_sts)






cwd = os.getcwd()